command/agent/job_endpoint_test.go

```diff
@@ -272,6 +272,38 @@ func TestHTTP_JobsRegister_Defaulting(t *testing.T) {
 	})
 }
 
+func TestHTTP_JobsParse(t *testing.T) {
+	t.Parallel()
+	httpTest(t, nil, func(s *TestAgent) {
+		buf := encodeReq(api.JobsParseRequest{JobHCL: mock.HCL()})
+		req, err := http.NewRequest("POST", "/v1/jobs/parse", buf)
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+
+		respW := httptest.NewRecorder()
+
+		obj, err := s.Server.JobsParseRequest(respW, req)
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		if obj == nil {
+			t.Fatal("response should not be nil")
+		}
+
+		job := obj.(*api.Job)
+		expected := mock.Job()
+		if job.Name == nil || *job.Name != expected.Name {
+			t.Fatalf("job name is '%s', expected '%s'", *job.Name, expected.Name)
+		}
+
+		if job.Datacenters == nil ||
+			job.Datacenters[0] != expected.Datacenters[0] {
+			t.Fatalf("job datacenters is '%s', expected '%s'",
+				job.Datacenters[0], expected.Datacenters[0])
+		}
```

Comment:
Would suggest checking all fields of the returned job against expected instead of just these three fields. 
We have a diff helper in https://github.com/hashicorp/nomad/blob/5c8b249b1c45e2b37f263ad617540ec9e2e8c2a0/nomad/structs/diff.go#L59 that we use in nomad plan and other places.

Reply:
https://github.com/hashicorp/nomad/blob/537b947612bbf65026380bdd027c8817671047ad/command/agent/job_endpoint.go#L547

Reply:
cautionary note - not sure if a direct translation like works okay when it comes to unset fields (esp if canonicalize is not called). so it may produce a diff complaining about empty list vs nil. I haven't tried it before, seem worth trying it and using it here.

Reply:
I spent some time looking at this and there are a few things:
* `Diff` errors if IDs don't match (mock.Job() generates a unique ID)
* Lots of unset fields yields a somewhat unuseful diff. 
* The `mock.HCL()` method doesn't quite mimic the `mock.Job()` fully.

At some point it feels like I'm testing more of the `jobspec.Parse` functionality than the api endpoint. Thoughts?

Reply:
That's fair. It doesn't seem worth it here to use diff for this test given the above.

Reply:
Thanks for the pointer though, I learned more about the structs package :smiley: